src/main/ApplicationImpl.cpp

```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+                    .getLastClosedLedgerHeader()
+                    .header.scpValue.closeTime);
+        }
+
+        return fmt::format(
+            FMT_STRING("Manually triggered a ledger close with sequence "
+                       "number {}"),
+            targetLedgerSeq);
     }
     else if (!mConfig.FORCE_SCP)
     {
         mHerder->setInSyncAndTriggerNextLedger();
-        return true;
+        return "Triggered a new consensus round";
+    }
+
+    throw std::invalid_argument(
+        "Set MANUAL_CLOSE=true in the stellar-core.cfg if you want to "
+        "close every ledger manually. Otherwise, run stellar-core "
+        "with --wait-for-consensus flag to close ledger once and "
+        "trigger consensus. Ensure NODE_IS_VALIDATOR is set to true.");
+}
+
+uint32_t
+ApplicationImpl::targetManualCloseLedgerSeqNum(
+    optional<uint32_t> const& userSpecifiedSeqNum)
+{
+    auto const startLedgerSeq = getLedgerManager().getLastClosedLedgerNum();
+
+    // The "scphistory" stores ledger sequence numbers as INTs.
+    if (startLedgerSeq >= std::numeric_limits<int32_t>::max())
+    {
+        throw std::invalid_argument(fmt::format(
+            FMT_STRING("Manually closed ledger sequence number ({}) can not be "
+                       "increased without overflowing database entry"),
+            startLedgerSeq));
+    }
+
+    auto const nextLedgerSeq = startLedgerSeq + 1;
+
+    if (userSpecifiedSeqNum)
+    {
+        if (*userSpecifiedSeqNum > std::numeric_limits<int32_t>::max())
+        {
+            throw std::invalid_argument(fmt::format(
+                FMT_STRING("Manual close ledger sequence number {} too large "
+                           "for database (maximum {})"),
+                *userSpecifiedSeqNum, std::numeric_limits<int32_t>::max()));
+        }
+
+        if (*userSpecifiedSeqNum <= startLedgerSeq)
+        {
+            throw std::invalid_argument(fmt::format(
+                FMT_STRING("Invalid manual close ledger sequence number {} "
+                           "(must exceed current sequence number {})"),
+                *userSpecifiedSeqNum, startLedgerSeq));
+        }
+    }
+
+    return userSpecifiedSeqNum ? *userSpecifiedSeqNum : nextLedgerSeq;
+}
+
+void
+ApplicationImpl::setManualCloseVirtualTime(
+    optional<TimePoint> const& userSpecifiedCloseTime)
+{
+    TimePoint const maxSystemTimePointSeconds =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            VirtualClock::system_time_point::duration::max())
+            .count();
+    TimePoint const maxTimeT = std::numeric_limits<std::time_t>::max();
+    TimePoint const maxCloseTime =
+        std::min(maxSystemTimePointSeconds, maxTimeT) -
+        Herder::MAX_TIME_SLIP_SECONDS.count();
+
+    TimePoint const lastCloseTime = getLedgerManager()
+                                        .getLastClosedLedgerHeader()
+                                        .header.scpValue.closeTime;
+    if (lastCloseTime >= maxCloseTime)
```

Comment:
Why do we care to compare `lastCloseTime` to the max? We should just check that the new time is within bounds and avoid duplicate code?
(this is true for both checks)

Reply:
The salient part of the `lastCloseTime` comparison is with `std::numeric_limits<TimePoint>::max()` -- maybe I should just test for equality with that.  If I don't, then the `lastCloseTime + 1` later might overflow, and that happens before we can check the new time.  With `secondsSinceEpoch`, that's making sure that `VirtualClock::to_time_t(systemNow)` won't overflow, which as Jon pointed out, could be undefined behavior.  So I might be able to change at least the first check to be against a different maximum (but I figured that as long as I was checking it, I may as well check it against the strictest thing that I could, as there would be no point continuing if it failed that check anyway).  Can I actually avoid either of these somehow without risking either of those overflows?

Reply:
That's because you wrote the code that way, if we use the current clock you won't do any arithmetic on time and can just enforce that `nextCloseTime > `lcl.ct`.

Reply:
> That's because you wrote the code that way, if we use the current clock you won't do any arithmetic on time and can just enforce that nextCloseTime > lcl.ct`.

Don't we also have to enforce that a manually-specified close time is greater than the _virtual_ system now, regardless of whether we use the system clock as a default, because `HerderImpl::triggerNextLedger()` is going to use the virtual time?

Reply:
I _think_ I've made the kind of rearrangement that you are requesting in 4932dab679e9f802c211ee35dd8d1c83655bb6e5.  Please let me know whether that is what you were looking for.

Reply:
yes things look good